test/integration/raft_test.lua

```diff
@@ -0,0 +1,323 @@
+                servers = {
+                    {instance_uuid = h.uuid('a', 'a', 1)},
+                },
+            },
+            {
+                alias = 'storage',
+                uuid = replicaset_uuid,
+                roles = {
+                    'vshard-storage',
+                    'test.roles.storage',
+                },
+                servers = {
+                    {
+                        instance_uuid = storage_1_uuid,
+                    },
+                    {
+                        instance_uuid = storage_2_uuid,
+                    },
+                    {
+                        instance_uuid = storage_3_uuid,
+                        env = {
+                            TARANTOOL_ELECTION_MODE = 'voter',
+                        },
+                    },
+                },
+            },
+        },
+    })
+    g.cluster:start()
+    set_failover_params{mode = 'raft'}
+end
+
+g.after_all = function()
+    g.cluster:stop()
+    fio.rmtree(g.cluster.datadir)
+end
+
+local function set_master(instance_name)
+    g.cluster:server(instance_name).net_box:eval[[
+        box.ctl.promote()
+    ]]
+end
+
+local function get_raft_info(alias)
+    return g.cluster:server(alias).net_box:eval([[
+        return box.info.election
+    ]])
+end
+
+local function kill_server(alias)
+    g.cluster:server(alias):stop()
+end
+
+local function start_server(alias)
+    g.cluster:server(alias):start()
+end
+
+local function get_master(uuid)
+    local response = g.cluster.main_server:graphql({
+        query = [[
+            query(
+                $uuid: String!
+            ){
+                replicasets(uuid: $uuid) {
+                    master { uuid }
+                    active_master { uuid }
+                }
+            }
+        ]],
+        variables = {uuid = uuid}
+    })
+    local replicasets = response.data.replicasets
+    t.assert_equals(#replicasets, 1)
+    local replicaset = replicasets[1]
+    return {replicaset.master.uuid, replicaset.active_master.uuid}
+end
+
+local function get_2pc_count()
+    local counts = {}
+    for _, server in ipairs(g.cluster.servers) do
+        table.insert(counts, server.net_box:eval([[
+            return _G['2pc_count']
+        ]]))
+    end
+    return counts
+end
+
+local function get_sharding_config()
+    local sharding = g.cluster:server('router-1'):eval[[
+        vars = require('cartridge.vars').new('cartridge.roles.vshard-router')
+        return vars.vshard_cfg['vshard-router/default'].sharding
+    ]]
+    return fun.iter(sharding):map(function(x, y)
+        return x, fun.iter(y.replicas):map(function(k, v)
+            return k, {master = v.master}
+        end):tomap()
+    end):tomap()
+end
+
+g.before_each(function()
+    h.retrying({}, function()
+        t.assert_equals(h.list_cluster_issues(g.cluster.main_server), {})
+    end)
+    h.retrying({}, function()
+        -- call box.ctl.promote on storage-1
+        set_master('storage-1')
```

Comment:
It's not guaranteed that 'storage-1' will be master after calling promote on it.
`box.ctl.promote` runs elections, and it is possible (though highly unlikely), that some other node wins them.
In case the node hasn't won the elections, promote will return an error.

So you may want to do something like `while not ok do ok = pcall(box.ctl.promote) end` to get the expected behaviour.

Or, as an alternative, set everyone except the promoted node to be `follower`.